### PR TITLE
Strong Semilattice of Semigroups

### PR DESCRIPTION
**Pull request by tomcontileslie**
_This PR supersedes #673; I've started a new one to squash the current log of ~30 commits and have easier write access to the PR (I was previously submitting PRs to Murray's branch)._

This PR is joint work with @james-d-mitchell, @MTWhyte and @reiniscirpons. It creates a new `StrongSemilatticeOfSemigroups` object initialised from a list of semigroups, a (meet-semilattice) digraph and a list of homomorphisms. Currently element enumeration, element comparison and multiplication is possible, which makes this a fully working **Semigroups** object, and it will be ready to merge as soon as it passes tests and review.

In future I will also be working off this branch to create additional features such as decomposition of certain types of completely regular semigroups into `StrongSemilatticeOfSemigroups` objects.

### CODE REVIEW

**Comment by tomcontileslie** [comment]
Just realised that there is inconsistency in the documentation calling an operation `UnderlyingStrongSemilatticeOfSemigroups` when it is actually implemented as `UnderlyingSemilatticeOfSemigroups`. I'll fix this in the next few days.

File: doc/semicons.xml [hunk 1]
```diff
@@ -299,3 +299,184 @@ true]]></Example>
...
+      If <A>SSS</A> is a strong semilattice of semigroups, this function
+      returns the underlying semilattice structure as a digraph.
+      Note that this may not be equal to the digraph passed as input
+      when <A>SSS</A> was created: rather, it is the reflexive transitive
+      closure of the input digraph.
+    </Description>
+  </ManSection>
+<#/GAPDoc>
+
+<#GAPDoc Label="UnderlyingStrongSemilatticeOfSemigroups">
```
Comment (tomcontileslie) [inline doc/semicons.xml:382]:
```suggestion
<#GAPDoc Label="UnderlyingSemilatticeOfSemigroups">
```

File: doc/semicons.xml [hunk 2]
```diff
@@ -299,3 +299,184 @@ true]]></Example>
...
+  <ManSection>
+    <Oper Name="UnderlyingStrongSemilatticeOfSemigroups" Arg="el"/>
```
Comment (tomcontileslie) [inline doc/semicons.xml:384]:
```suggestion
    <Oper Name="UnderlyingSemilatticeOfSemigroups" Arg="el"/>
```

**Review by james-d-mitchell — APPROVED** [review]
I'm happy with this apart from some minor comments. One other comment is: have you checked that the documentation looks ok when viewed as text only?

File: doc/z-chap09.xml [hunk 3]
```diff
@@ -32,4 +32,38 @@
...
+    in the semigroup <M>S_c</M>, corresponding to the meet <M>c</M> of
+    <M>a, b \in Y</M>. The exact element of <M>S_c</M> equal to the product
+    is obtained using the homomorphisms of the SSS: <M>xy = (x f_{ac})
+    (y f_{bc})</M>.
+
+    <#Include Label = "StrongSemilatticeOfSemigroups">
+    <#Include Label = "SSSE">
+    <#Include Label = "IsSSSE">
+    <#Include Label = "IsStrongSemilatticeOfSemigroups">
+    <#Include Label = "StrongSemilatticeOfSemigroups">
```
Comment (james-d-mitchell) [inline doc/z-chap09.xml:60]:
```suggestion
```

**Comment by james-d-mitchell** [comment]
I think I'm ready to merge this now, any reason not to @tomcontileslie ?

**Comment by tomcontileslie** [comment]
@james-d-mitchell I've just updated the tests since your review updated error messages printed to the console in the standard test suite. All clear for merge on my end.

**Comment by wilfwilson** [comment]
When this PR is ready to be merged, I suggest doing a squash-and-merge to tidy up the commit history (see https://github.com/semigroups/Semigroups/pull/752#issuecomment-789145063). Or, if @tomcontileslie is confident and knowledgeable enough with `git` to rewrite the history of this branch via an interactive rebase and then force push a cleaned up version, then go ahead.

**Comment by tomcontileslie** [comment]
@wilfwilson I've tidied up the history into three commits that show broad chronological progression (and each commit should pass tests). Let me know whether that works for you, I can always squash further.

**Comment by wilfwilson** [comment]
Nice one!

**Pull request merged**